plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2953,23 +3009,24 @@ bool producer_plugin_impl::push_read_only_transaction(
    }
 
    try {
-      const auto block_deadline = calculate_block_deadline( chain.pending_block_time() );
       auto start = fc::time_point::now();
-      auto time_used_in_read_window_us = ( start - read_window_start_time );
-      if ( time_used_in_read_window_us >= _ro_read_window_time_us ) {
+      if ( start >= _ro_window_deadline ) {
          // already passed read-window deadline, try next round
          return true;
       }
 
-      auto remaining_time_in_read_window_us = _ro_read_window_time_us - time_used_in_read_window_us;
-      // Ensure the trx to finish by the end of read-window.
-      auto window_deadline = std::min( start + remaining_time_in_read_window_us, block_deadline );
+      // when executing on the main thread while in the write window, need to switch db mode to read only
+      // _ro_in_read_only_mode can only be false if running on main thread as it is only modified from the main thread
+      const bool switch_read_mode = !_ro_in_read_only_mode;
+      auto db_read_only_mode_guard = fc::make_scoped_exit([&]{
+         if( switch_read_mode )
+            chain.unset_db_read_only_mode();
+      });
+      if( switch_read_mode )
+         chain.set_db_read_only_mode();
```

Comment:
Now those 3 statements will be always executed, even when read-only threads are enabled.

Reply:
Yes. Do you think that is a problem?